# Language detection: Spacy vs Pycld2

This notebook will compare the language detection between Spacy and Pycld2

In [1]:
import pandas as pd
from colorama import Fore

In [2]:
import spacy
from spacy_langdetect import LanguageDetector

In [3]:
import pycld2 as cld2

In [4]:
pathori = "../data/original"
pathdest = "../data/preprocessed/"
pathmodel = "../data/models/"
debug = 1

In [12]:
def load_data():
    languages = {"ca":"ca_fasttext_wiki", "es":"es_core_news_sm", 
                 "en":"en_core_web_sm", "xx":"" }

    data_total = pd.DataFrame(columns=["Comentari","TipusIncidencia","Idioma"])

    for language in languages:
        model = languages[language]
        file = "comentaris_" + language + ".csv"
        if (debug >= 1):
            print ("language: ", language)
            print ("model: ", model)
            print ("file:", file)

        # Load data from file

        data_csv = pd.read_csv(pathdest + file)
        data_lang = data_csv[["Comentari","TipusIncidencia","Idioma"]]
        if (debug >= 2):
            display (data_lang.sample(5))
        if (debug >= 1):
            print ("rows:", data_lang.shape[0])
            print ("")

        data_total = pd.concat([data_total, data_lang])
        
    data_total = data_total[data_total["Comentari"].notnull()]
    data_total = data_total.reset_index()
    if (debug >= 1):
        print ("data_total: ", data_total.shape[0], " rows")

    if (debug >= 2):
        display (data_total.sample(5))
        
    return data_total

In [6]:
def detect_language_pycld2(comment):
    isReliable, textBytesFound, details = cld2.detect(comment.encode('utf-8', 'replace'), 
                                                    isPlainText = True, bestEffort = True, returnVectors=False)
    
    return details[0][1]

In [13]:
data = load_data()
if (debug >= 1):
    display(data)
    
data = detect_language_rows(data)
if (debug >= 1):
    display(data)

language:  ca
model:  ca_fasttext_wiki
file: comentaris_ca.csv
rows: 5999

language:  es
model:  es_core_news_sm
file: comentaris_es.csv
rows: 1345

language:  en
model:  en_core_web_sm
file: comentaris_en.csv
rows: 330

language:  xx
model:  
file: comentaris_xx.csv
rows: 185

data_total:  7855  rows


index                                          Comentari  \
0         0  Els continguts d'aquesta assignatura ja els ha...   
1         1  L'actitud de la professora em resulta molt mot...   
2         2  Materials molt interessants i bones orientacio...   
3         3  Ha estat una exigència molt forta de treball: ...   
4         4  Molta satisfacció amb la transparència i l'ant...   
...     ...                                                ...   
7850    180                        Un exemple de bon professor   
7851    181                                     No sé quién es   
7852    182                                          PROFE TOP   
7853    183                                  Sense paraules...   
7854    184                                       No se qui és   

                          TipusIncidencia Idioma  
0                                     NaN     ca  
1                                     NaN     ca  
2                                     NaN     ca  
3                                     NaN     ca  
4                                     NaN     ca  
...                                   ...    ...  
7850                                  NaN     fr  
7851  No ha impartit classe a aquest grup     fr  
7852                                  NaN     vi  
7853                                  NaN     fr  
7854  No ha impartit classe a aquest grup     fr  

[7855 rows x 4 columns]

In [14]:
#data["IdiomaPycld2"] = data.apply(lambda row: detect_language_pycld2(row["Comentari"]), axis=1)    
#data["IdiomaPycld2"] = detect_language_pycld2(data["Comentari"])

# Obtain IdiomaPycld2
def detect_language_rows(data):
    for x in range(data.shape[0]):
        comment = data.loc[x]["Comentari"]   
        idiomaPycld2 = detect_language_pycld2(comment)
        data.loc[x, "IdiomaPycld2"] = idiomaPycld2

    return data

index                                          Comentari  \
0         0  Els continguts d'aquesta assignatura ja els ha...   
1         1  L'actitud de la professora em resulta molt mot...   
2         2  Materials molt interessants i bones orientacio...   
3         3  Ha estat una exigència molt forta de treball: ...   
4         4  Molta satisfacció amb la transparència i l'ant...   
...     ...                                                ...   
7850    180                        Un exemple de bon professor   
7851    181                                     No sé quién es   
7852    182                                          PROFE TOP   
7853    183                                  Sense paraules...   
7854    184                                       No se qui és   

                          TipusIncidencia Idioma IdiomaPycld2  
0                                     NaN     ca           ca  
1                                     NaN     ca           pt  
2                                     NaN     ca           ca  
3                                     NaN     ca           ca  
4                                     NaN     ca           ca  
...                                   ...    ...          ...  
7850                                  NaN     fr           en  
7851  No ha impartit classe a aquest grup     fr           es  
7852                                  NaN     vi           en  
7853                                  NaN     fr           ca  
7854  No ha impartit classe a aquest grup     fr           hu  

[7855 rows x 5 columns]

In [9]:
#   Distribution by Idioma        
data_group = data[["Idioma","Comentari"]].groupby(["Idioma"]).count()
if debug:
    print ("Idioma Spacy:")
    display (data_group.sort_values("Comentari", ascending=False).head(10))

Idioma Spacy:


Comentari
Idioma            
ca            5991
es            1345
en             330
it              42
fr              42
pt              38
UNKNOWN         10
de               8
hu               6
nl               6

In [10]:
#   Distribution by IdiomaPycld2        
data_group = data[["IdiomaPycld2","Comentari"]].groupby(["IdiomaPycld2"]).count()
if debug:
    print ("Idioma Pycld2:")
    display (data_group.sort_values("Comentari", ascending=False).head(10))

Idioma Pycld2:


Comentari
IdiomaPycld2           
ca                 5962
es                 1343
en                  426
fy                   25
tk                   14
un                   13
hu                   13
oc                    8
pt                    7
gl                    6

In [11]:
#   Comparativa idiomas        
data_group = data[["Idioma","IdiomaPycld2","Comentari"]].groupby(["Idioma","IdiomaPycld2"]).count().reset_index()
display (data_group.sort_values("Comentari", ascending=False).head(10))
data_pivot = data_group.pivot(index="Idioma", columns="IdiomaPycld2", values="Comentari")
if debug:
    display (data_pivot)

Idioma IdiomaPycld2  Comentari
3      ca           ca       5815
31     es           es       1297
24     en           en        239
5      ca           en        135
23     en           ca         61
53     it           ca         23
30     es           en         20
74     pt           es         19
41     fr           ca         16
29     es           ca         16

IdiomaPycld2   af      ca  crs   cs   cy   de     en      es   fi   fr  ...  \
Idioma                                                                  ...   
UNKNOWN       NaN     NaN  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
af            1.0     1.0  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
ca            NaN  5815.0  NaN  NaN  NaN  1.0  135.0    11.0  NaN  1.0  ...   
cy            NaN     NaN  NaN  NaN  NaN  NaN    NaN     1.0  NaN  NaN  ...   
de            NaN     3.0  NaN  NaN  NaN  NaN    1.0     4.0  NaN  NaN  ...   
en            NaN    61.0  NaN  NaN  NaN  NaN  239.0     5.0  NaN  NaN  ...   
es            NaN    16.0  NaN  NaN  NaN  NaN   20.0  1297.0  NaN  NaN  ...   
et            NaN     1.0  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   
fi            NaN     1.0  NaN  NaN  NaN  NaN    NaN     NaN  1.0  NaN  ...   
fr            NaN    16.0  NaN  NaN  NaN  NaN   10.0     3.0  NaN  3.0  ...   
hu            NaN     6.0  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
id            NaN     2.0  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   
it            1.0    23.0  1.0  NaN  NaN  NaN    4.0     3.0  NaN  NaN  ...   
lt            NaN     NaN  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
nl            NaN     2.0  NaN  NaN  NaN  NaN    2.0     NaN  NaN  NaN  ...   
no            NaN     1.0  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   
pl            NaN     NaN  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
pt            NaN     6.0  NaN  3.0  1.0  NaN    6.0    19.0  1.0  NaN  ...   
ro            NaN     NaN  NaN  NaN  NaN  NaN    3.0     NaN  NaN  NaN  ...   
sl            NaN     1.0  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
so            NaN     NaN  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   
sq            NaN     NaN  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
sv            NaN     2.0  NaN  NaN  NaN  NaN    NaN     NaN  NaN  NaN  ...   
sw            NaN     NaN  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   
vi            NaN     1.0  NaN  NaN  NaN  NaN    1.0     NaN  NaN  NaN  ...   

IdiomaPycld2   lb   ms   oc   pt   sk    tk   tn    un   xh   yo  
Idioma                                                            
UNKNOWN       NaN  NaN  NaN  NaN  NaN   NaN  NaN  10.0  NaN  NaN  
af            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
ca            1.0  1.0  4.0  6.0  1.0   NaN  NaN   2.0  NaN  1.0  
cy            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
de            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
en            NaN  NaN  NaN  NaN  NaN  13.0  NaN   NaN  NaN  NaN  
es            NaN  NaN  2.0  NaN  1.0   NaN  NaN   NaN  NaN  NaN  
et            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
fi            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
fr            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  5.0  
hu            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
id            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
it            NaN  NaN  1.0  NaN  NaN   NaN  1.0   NaN  NaN  NaN  
lt            NaN  NaN  NaN  NaN  NaN   NaN  NaN   1.0  NaN  NaN  
nl            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
no            NaN  NaN  NaN  NaN  NaN   1.0  NaN   NaN  NaN  NaN  
pl            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
pt            NaN  NaN  NaN  1.0  NaN   NaN  NaN   NaN  NaN  NaN  
ro            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
sl            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
so            NaN  1.0  NaN  NaN  NaN   NaN  NaN   NaN  1.0  NaN  
sq            NaN  NaN  1.0  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
sv            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
sw            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  
vi            NaN  NaN  NaN  NaN  NaN   NaN  NaN   NaN  NaN  NaN  

[25 rows x 28 columns]

In [64]:
data_dif = data[data["Idioma"] != data["IdiomaPycld2"]]
display (data_dif)
print (data_dif.shape[0])

index                                          Comentari Idioma  \
1         1  L'actitud de la professora em resulta molt mot...     ca   
53       53                              M'ha sembla correcta.     ca   
54       54                                         Magnífica.     ca   
63       63                          Falten exercicis per fer.     ca   
72       72                                  És un incompetent     ca   
...     ...                                                ...    ...   
7850    172                                     No sé quién es     fr   
7851    173                                          PROFE TOP     vi   
7852    174  Les professores a vegades et diuen coses contr...     fr   
7853    175                                  Sense paraules...     fr   
7854    176                         No he demanat cap tutoria.     it   

     idioma_pycld2 IdiomaPycld2  
1                            pt  
53                           ie  
54                           es  
63                           de  
72                           hu  
...            ...          ...  
7850                         es  
7851                         en  
7852                         ca  
7853                         ca  
7854                         en  

[500 rows x 5 columns]

500


In [99]:
data_dif = data[(data["Idioma"] == "ca") & (data["IdiomaPycld2"] != "ca") & (data["IdiomaPycld2"] != "es") & (data["IdiomaPycld2"] != "en")]
display (data_dif)
print (data_dif.shape[0], " rows")

index                                          Comentari Idioma  \
1         1  L'actitud de la professora em resulta molt mot...     ca   
53       53                              M'ha sembla correcta.     ca   
63       63                          Falten exercicis per fer.     ca   
72       72                                  És un incompetent     ca   
111     111             Per venir a vendre la empresa només...     ca   
201     201  Teoria rapida i temps de practica a classe. Co...     ca   
250     250                         És un professor correcte.      ca   
365     365   No puc evaluar-lo, no estava quan va fer classe.     ca   
367     367     Sempre disposat a ajudar en el que necessitis.     ca   
402     402                                     És una crack!!     ca   
521     521                               Classe massa teòrica     ca   
653     653                  Un expert i crack en la matèria!      ca   
684     684                         Manera d'explicar pessada.     ca   
788     788                                    No s'explica bé     ca   
918     918  Molt content en com sha portat lassignatura de...     ca   
1097   1097  Ens ensenyen tecnomogies obsoletes, pero ens e...     ca   
1198   1199  Alguns cops penso que faltava preparació prèvi...     ca   
1509   1510  El funcionament del primer parcial durant la q...     ca   
1680   1681                     És una excel·lent professora.      ca   
1730   1731                            És una gran professora.     ca   
1800   1801  NO val per res, nomes pensa en putejar el pers...     ca   
2117   2118                              És una professora 10.     ca   
3020   3021                             Es una Bona professora     ca   
3296   3298  L'Enric ha estat un professor que s'ha adaptat...     ca   
4833   4835                     És un 10 de professora!!!!!!!!     ca   
5139   5143                            És una bona professora.     ca   
5218   5222                        És una docent exemplar.\r\n     ca   
5693   5697                           ´Metòdic i bon professor     ca   
5870   5874                                       No se qui és     ca   
5901   5905                                  Res a dir. Genial     ca   
5927   5931                       Un 10 de persona i professor     ca   

     idioma_pycld2 IdiomaPycld2  
1                            pt  
53                           ie  
63                           de  
72                           hu  
111                          fr  
201                          un  
250                          hu  
365                          pt  
367                          pt  
402                          hu  
521                          un  
653                          oc  
684                          pt  
788                          yo  
918                          sk  
1097                         ht  
1198                         pt  
1509                         oc  
1680                         hu  
1730                         hu  
1800                         ms  
2117                         hu  
3020                         fy  
3296                         oc  
4833                         hu  
5139                         hu  
5218                         jw  
5693                         oc  
5870                         hu  
5901                         lb  
5927                         fy

31  rows


In [102]:
data_dif = data[(data["Idioma"] == "es") & (data["IdiomaPycld2"] != "es") & (data["IdiomaPycld2"] != "ca") 
                & (data["IdiomaPycld2"] != "en")]
display (data_dif)
print (data_dif.shape[0], " rows")

index                                          Comentari Idioma  \
6026     21    Materiales estupendos para disfrutar y aprender     es   
6168    163                No vimos a en clase a este docente      es   
6256    251                              No vimos clase con él     es   
6257    252                            No vimos clase con ella     es   
6334    329              Excelente para la formación de Master     es   
6345    340                 mala organización de la asignatura     es   
6410    405     Los testes de teoria son tediosos y dificiles.     es   
6606    601     De dificil acceso y ausente durante el curso.      es   
6770    765  No fica examen del teoria i fica coses que no ...     es   
6818    813                                        Bona profe.     es   
7002    997  Pese a que esta asignatura debe de ser presenc...     es   
7117   1112  Los métodos utilizados por Ines son los adecua...     es   
7188   1183            Trobo que no esta gaire ben plantejada      es   

     idioma_pycld2 IdiomaPycld2  
6026                         id  
6168                         gl  
6256                         gl  
6257                         gl  
6334                         gl  
6345                         sk  
6410                         oc  
6606                         gl  
6770                         pt  
6818                         fy  
7002                         gl  
7117                         id  
7188                         oc

13  rows


In [105]:
data_dif = data[(data["Idioma"] == "en") & (data["IdiomaPycld2"] != "en") & (data["IdiomaPycld2"] != "ca")
                & (data["IdiomaPycld2"] != "es")]
display (data_dif)
print (data_dif.shape[0], " rows")

index                                 Comentari Idioma idioma_pycld2  \
7349      0                             Gran profesor     en                 
7351      2                           No l'hem tingut     en                 
7359     10                           No l’hem tingut     en                 
7360     11                           No l'hem tingut     en                 
7362     13                           No l’hem tingut     en                 
7368     19                   És mlt bon professor ;)     en                 
7442     93                           No l'hem tingut     en                 
7451    102                           No l’hem tingut     en                 
7458    109                           No l’hem tingut     en                 
7500    151                          No l'hem tingut.     en                 
7552    203                           No l'hem tingut     en                 
7584    235                           No l'hem tingut     en                 
7600    251       Professor molt útil, bon professor.     en                 
7607    258                          No l’hem tingut.     en                 
7609    260                            Bon profesor.      en                 
7610    261                        Un bon professor.      en                 
7613    264          No l’hem tingut a fisiologia II.     en                 
7616    267                        Professorat inepte     en                 
7626    277                           Bona professora     en                 
7634    285  No hem tingut a l'Olga com a professora.     en                 
7638    289                        Un bon professor.      en                 
7646    297                            Bon professor.     en                 
7654    305                            Gran profesor.     en                 
7655    306                             Bon professor     en                 
7658    309                             Bon professor     en                 

     IdiomaPycld2  
7349           fy  
7351           tk  
7359           tk  
7360           tk  
7362           tk  
7368           hu  
7442           tk  
7451           tk  
7458           tk  
7500           tk  
7552           tk  
7584           tk  
7600           fy  
7607           tk  
7609           fy  
7610           fy  
7613           tk  
7616           fy  
7626           fy  
7634           tk  
7638           fy  
7646           fy  
7654           fy  
7655           fy  
7658           fy

25  rows


In [110]:
data_dif = data[(data["Idioma"] != "ca") & (data["Idioma"] != "es") & (data["Idioma"] != "en")
                & (data["IdiomaPycld2"] != "ca") & (data["IdiomaPycld2"] != "es") & (data["IdiomaPycld2"] != "en")]
display (data_dif)
print (data_dif.shape[0], " rows")

index                                          Comentari   Idioma  \
7687      9  `............````.`......`.`````.................       so   
7689     11                                          ¯\_(?)_/¯  UNKNOWN   
7692     14                              No em crida l'atenció       pt   
7697     19                                 Excelente docente        pt   
7698     20                                 Excelente docente.       pt   
7707     29                                  Excelente docente       pt   
7710     32                               Excelente profesor.        ro   
7717     39                                Excelente profesora       ro   
7718     40                                   No tuvimos clase       lt   
7719     41  Comença les classes a les 16.10h en compte de ...       fr   
7720     42                                       Puntua baix.       id   
7723     45                                       No sé qui és       fr   
7725     47                      No te se sembla a la realitat       it   
7727     49                                   Molt bon docent.       nl   
7728     50                                       No sé qui és       fr   
7730     52                                                Bé.       fr   
7731     53  Comença les classes a les 16.10h en compte de ...       fr   
7738     60                                     No hem fet res       no   
7742     64                                       <3 a Haskell       no   
7743     65                                     No puc jutjar.       sq   
7745     67                                   Molt bon docent.       nl   
7747     69                                            Tot dit       af   
7748     70                           Tota una bona professora       it   
7750     72                                               \r\n  UNKNOWN   
7753     75                                               Util       sw   
7756     78                                               \r\n  UNKNOWN   
7758     80                                                  0  UNKNOWN   
7762     84              bona professora, propera i motivadora       it   
7763     85                   bona professora, propera i clara       it   
7766     88                                                  -  UNKNOWN   
7770     92                               La professora un 10.       it   
7778    100                                                 Bé       fr   
7779    101                                             Bé\r\n       fr   
7781    103                     Bona persona. i profesora \r\n       it   
7782    104                               Si li fas la pilota        it   
7785    107                               És un bon professor.       fr   
7786    108                                                 Bé       fr   
7787    109                                                 Bé       fr   
7788    110                                    Bona professora       it   
7789    111                                   un bon professor       fr   
7790    112                                       Cantar massa       pt   
7794    116                               Se li agafa carinyo.       so   
7795    117                         Ester excelente profesora.       pt   
7796    118                        Ester excelente  profesora.       pt   
7801    123                                Professora nefasta!       it   
7814    136                                                ---  UNKNOWN   
7816    138                                  S'ho creu massa!        pt   
7818    140                                                ---  UNKNOWN   
7823    145                                          Magnifica       it   
7835    157                                                  -  UNKNOWN   
7841    163                                               \r\n  UNKNOWN   
7844    166                                               \r\n  UNKNOWN   
7845    167                    

55  rows


In [17]:
# Registro con idioma diferente agrupados por tipo de incidencia

data_dif = data[data["Idioma"] != data["IdiomaPycld2"]]
data_incid = data_dif[data_dif["TipusIncidencia"].notnull()]
data_group = data_incid[["TipusIncidencia","Comentari"]].groupby(["TipusIncidencia"]).count()
data_group

Comentari
TipusIncidencia                               
Comentari de professor                       2
Comentari excel·lent                         1
Comentari ofensiu                            2
Comentari problemàtic                        5
Exclamacions o emoticones excessius          1
No ha impartit classe a aquest grup         58

Nivel general de coincidencia:
- Textos con igual idioma: 7.351 / 7855 = 93,58 %
- Textos con idioma diferente: 504 registros. Candidatos a recuperarse.

Revisión idioma catalán:
- Los textos detectados ca (spacy) - en (pycld2) són en realidad en catalán (141 casos)
- Los textos detectados ca (spacy) - es (pycld2) són en realidad en catalán (11 casos)
- Los textos detectados ca (spacy) - xx (pycld2) són en realidad en catalán (31 casos)
Conclusión: En general la detección de catalán de Spacy es correcta

Revisión idioma castellano:
- Los textos detectados es (spacy) - ca (pycld2) són en realidad en catalán (16 casos)
- Los textos detectados es (spacy) - en (pycld2) són en realidad en castellano (19 casos)
- Los textos detectados es (spacy) - xx (pycld2) són en realidad en castellano (13 casos)
Conclusión: En geneal la detección del castellano en Spacy es correcta, salvo para el idioma catalán, que se podria mejorar mediante pycld2 

Revisión idioma inglés:
- Los textos detectados en (spacy) - ca (pycld2) són en realidad en catalán (60 casos)
- Los textos detectados en (spacy) - es (pycld2) són en realidad en castellano (8 casos)
- Los textos detectados en (spacy) - xx (pycld2) són en realidad en catalán (25 casos)
Conclusión: Para el castellano y el catalán la detección de pycld2 es mejor y en el caso de otras lenguas se debería hacer una asignación directa al catalán.

Revisión otras lenguas (xx representa las lengua que no son ca, es, en):
- Los textos detectados xx (spacy) - ca (pycld2) són en realidad en catalán (68 casos)
- Los textos detectados xx (spacy) - es (pycld2) són en realidad en castellano (28 casos)
- Los textos detectados xx (spacy) - en (pycld2) són en realidad en catalán (30 casos)
- Los textos detectados xx (spacy) - xx (pycld2) són casi todos catalán y el resto castellano (55 casos)

Conclusión: Para el castellano y el catalán la detección de pycld2 es mejor y en el caso de otras lenguas se debería hacer una asignación directa al catalán.

Agrupación por tipos de incidencia:
- De los 505 registro recuperables 69 son incidencias.
- La mayoría de las incidencias (58 registros) son de tipo 'No ha impartido clase'.
- En todo el conjunto de datos hay 440 incidencias, 37 de tipo 'No ha impartido clase'.
